File: app/src/main/java/com/sjn/taggingplayer/db/dao/DeviceDao.java
package com.sjn.taggingplayer.db.dao;

import com.sjn.taggingplayer.db.Device;

import io.realm.Realm;

public class DeviceDao extends BaseDao {

    private static DeviceDao sInstance;

    public static DeviceDao getInstance() {
        if (sInstance == null) {
            sInstance = new DeviceDao();
        }
        return sInstance;
    }

    public Device findOrCreate(Realm realm, Device rawDevice) {
        Device device = realm.where(Device.class).equalTo("mModel", rawDevice.getModel()).equalTo("mOs", rawDevice.getOs()).findFirst();
        if (device == null) {
            device = realm.copyToRealm(rawDevice);
            device.setId(getAutoIncrementId(realm, Device.class));
        }
        return device;
    }

    public Device newStandalone() {
        return new Device();
    }

}

File: app/src/main/java/com/sjn/stamp/model/dao/DeviceDao.kt
package com.sjn.stamp.model.dao

import com.sjn.stamp.model.Device
import io.realm.Realm

object DeviceDao : BaseDao<Device>() {

    fun findOrCreate(realm: Realm): Device {
        val device = Device()
        return findOrCreate(realm, device.model, device.os)
    }

    fun findOrCreate(realm: Realm, model: String, os: String): Device {
        var device: Device? = realm.where(Device::class.java).equalTo("model", model).equalTo("os", os).findFirst()
        if (device == null) {
            return if (realm.isInTransaction) {
                realm.createObject(Device::class.java, CategoryStampDao.getAutoIncrementId(realm)).apply {
                    this.model = model
                    this.os = os
                }
            } else {
                realm.beginTransaction()
                device = realm.createObject(Device::class.java, CategoryStampDao.getAutoIncrementId(realm)).apply {
                    this.model = model
                    this.os = os
                }
                realm.commitTransaction()
                device
            }
        }
        return device
    }

}
